### DISCUSSION TITLE
Using `Linq2Acad` in more complex situations

### DISCUSSION THREAD

**Discussion by stenci** [wtertinek/Linq2Acad#39] [Q&A] [answered]
The examples show how to use `Linq2Acad` in simple situations, where all the code fits inside a small `using` block.

I tried to use it in the real world, replacing all the horrible boiler code that is usually required, and I can't find a nice way to use it when the database is accessed deep down in the call hierarchy.

For example I have a `MaterialKey` class with the method `MaterialKey.CreateBlock`. The class is used by the `SettingManagement` class, which is used by other classes. In my current implementation, `CreateBlock` uses `TopTransaction` and doesn't care where the transaction was started, it just assumes that the transaction exists and can be used.

I could leave `CreateBlock` as is, relying that `TopTransaction` will behave correctly, and it will work because it doesn't care whether the transaction was created by `Linq2Acad` or in some other way. But I wouldn't be able to use all the goodies that come with `Linq2Acad`.

Should I do the same with `Linq2Cad` and create a global `TopAcadDatabase` that is assumed to be initialized somewhere in the call stack, just like now `TopTransaction` is assumed to return a valid transaction?

**Comment by benkoshy** [comment]
Post the code? It is very difficult to reason with abstract code. in addition it may help solve the X-Y problem.

**Comment by wtertinek** [comment, marked as answer]
Hi stenci,

I think you have several possibilities:

1) The simplest (and probably cleanest) way to handle the issue would be to create a single instance of `AcadDatabase` (within a using statement) and pass it along your class hierarchy or call chain (within the body of the using statement). Doing so, you do not rely on global state and the state (`AcadDatabase`) is passed along / injected. This requires that all the actions you want to perform on the database are done using the `AcadDatabase` object. 

```csharp
public void CreateBlock(AcadDatabase acadDb)
{
  
}
```

2) Another way would be to start a transaction and pass the transaction object along your call chain. Doing so you control the life cycle of the transaction, preventing transaction related issues. This gives you the flexibility to do whatever you want, use `AcadDatabase` or not. To use the transaction with the `AcadDatabase` you can use an overload of `AcadDatabase.Active`:

```csharp
AcadDatabase.Active(Transaction transaction, bool commitTransaction, bool disposeTransaction)
```

You can use it like:

```csharp
public void CreateBlock(Transaction tr)
{
  var acadDb = AcadDatabase.Active(tr, false, false);
}
```

3) An alternative to 2) would be the way you mentioned, to simply rely on the `TopTransaction`, assuming it exists and can be used:

```csharp
public void CreateBlock()
{
  var db = GetDatabase();
  var acadDb = AcadDatabase.Active(db.TransactionManager.TopTransaction, false, false);
}
```

I hope that helps clarifying your issues.

**Comment by stenci** [comment]
The clean way would be good when starting from scratch, but I don't want to refactor everything because tomorrow that everything could call a method that requires the argument `acadDb`.

I like the `AcadDatabase.Active(tr, false, false)` because it will allow me to use Linq2Acad with existing code and starting to get familiar with it.

**Comment by stenci** [comment]
If I use `AcadDatabase.Active(db.TransactionManager.TopTransaction, false, false)`, I am working with an `IDisposable`, so I should make sure I dispose it.

What is the best practice in this case?
- In this case I don't really need to dispose, because I am explicitly telling it not to dispose, which _I know_ it means there is no need to dispose
- I should never rely on what _I know_ when I use a package that is not mine. Instead I should follow the guidelines, that is always make sure every `IDisposable` is disposed, and leave it to it to decide what to do at dispose time.

**Reply by wtertinek** [reply]
This is a very good observation. Absolutely correct. I haven't thought about it this way, but design of the `AcadDatabase.Active(db.TransactionManager.TopTransaction, false, false)` is actually flawed. If we want the object not to be disposed, the the API should not event provide the way to do it.

Currently the method signature is

```csharp
public static AcadDatabase Active(Transaction transaction, bool commitTransaction, bool disposeTransaction)
{
}
```

Instead, something like a `Reuse` method would be much better which returns an instance of a new class, something like `NotDisposableAcadDatabase`:

```csharp
public static NotDisposableAcadDatabase Reuse(Transaction transaction)
{
}
```

Here the API leaves no room for interpretation, since `NotDisposableAcadDatabase` does not implement `IDisposable` and there is no need for guidelines.

So

```csharp
using (var db = AcadDatabase.Reuse(tr))
{
}
```

would not even be possible and yield an error.

If you want to further develop the idea and maybe also want to implement it, don'T hesitate to provide a pull request.

**Reply by stenci** [reply]
I understand what you mean by "the current design of `Active` is flawed", why implementing a `NotDisposableAcadDatabase` would fix that flaw, and I might give it a try.

But now I see that:
- On one hand you consider a flaw an API design that asks whether to dispose or not an `IDisposable` and you suggest to create non disposable classes instead. This makes me think "all those guidelines about always disposing `IDisposable`s" are correct.
- On the other hand [you say](https://github.com/wtertinek/Linq2Acad/discussions/41#discussioncomment-8499310) it's OK not to dispose of some `IDisposable` when _you know_ it's OK not to.

So, here I am, back at the starting point, trying to figure out what's the right thing to do with this weird C#.
I have started to manage my `IDisposable`s using [this module](https://forums.autodesk.com/t5/net/how-do-you-dispose-lazily-loaded-created-objects/m-p/12546982/highlight/true#M80493), and things seem to be more manageable. That `AutoDispose` is really handy. But I'm still confused.

Thank you for your help, it will take some time, but I will get there. 😄

**Reply by wtertinek** [reply]
OK, now I know what you actually want to know 😄 Concerning the life cycle of AutoCAD objects I go with the [recomendation](https://www.keanw.com/2008/06/cleaning-up-aft.html) on Through the Interface:

1. Temporary objects - such as those provided by Autodesk.AutoCAD.Geometry - which are never Database-resident
Safe to be disposed of either **"manually"** (by your own code) or **"automatically"** (by the .NET garbage collector)
2. Temporary objects with the potential to be database-resident but which never actually get added to a Database
**Must be disposed of manually**
3. Database-resident objects added/accessed via a Transaction
**Transaction is able to dispose of them automatically**. Make sure you wrap your use of Transactions in using blocks or call Dispose() on them when you're done.

Since Linq2Acad only deals with database-resident objects (3.), there's no need to bother with `IDisposable`s.

From the use case described in your link, I understand that this relates to 2. (temporary database objects). So calling `Dispose()` would be required. But I actually don't really understand the use case: why would I create temporary objects but treat them as long-lived objects and manage their life cycle outside of the database?

But maybe I'm missing something, I've almost always worked with database-resident objects.

**Reply by stenci** [reply]
> OK, now I know what you actually want to know

In the past I have created little C# desktop tools, I played with Inventor and Revit addins, enough to decide not to continue and switch to AutoCAD, but this AutoCAD addin is my first large C# project. One day I noticed warnings in the Debug window asking me if I had forgotten to call `Dispose` on some `IDisposable`s, so I started digging into what's the best practice, and I keep bumping into 3 walls: one that says you must always dispose all the `IDisposable`s you deal with, one that says in some cases it's a waste of time and readability, and one that says there are objects that if you Dispose of them, you break stuff.

By now I understand that there are no best practices, you need to know what you are doing. 
I am sad.
😢 

---

Rant of the day (please ignore)

I started working in C in the eighties, when there were no auto formatters (that I knew of), and every day I cursed a language that requires variable type declaration, memory allocation and curly braces. I had to manually indent the code because it helped me understanding the structure of the code, that should have been enough for the compiler to understand it too, but no, I had to put those ugly curly braces. I could have said `x = 0.0`, but no, I had to say that `x` was a `double` too. 

After years seeing pseudo code that explained the logic followed by real code with curly braces and type declaration and other useless stuff, one day I saw a Wikipedia page with nice pseudo code that was actually working: it was the day I met Python, where you don't need to say that "abc" is a string, and you don't need to declare the code structure twice, once by indenting and once with curly braces. I never looked back. Since that day I've done only my job, taking care of the logic, and left the rest to Python.

Now I'm working on an AutoCAD addin and I'm forced to go back to the Mesozoic, I can't believe that in 2024 still there are languages where you need to tell the compiler that a number is a number and a string is a string, you need to use curly braces because indenting is not enough, and you even need to dispose of stuff.

I know, I know, Python is slow, and Python has the `with`, equivalent of the `using`. But Python is also used very efficiently with LLM and tons of other libraries. And the very reason why those powerful libraries are so popular today is that they don't require you to waste time declaring and formatting and disposing, all things the IDE / compiler should do for you.

Over the years I've seen plenty of people asking how to speed up C# code, which was too complex to speed up, while using Python things were much faster. Because you see what the code does instead of seeing curly braces, declarations, dispositions, etc.

Your library did the same for me: a function that compares the content of two drawings was very slow because it opened the reference drawing. I knew I could have read its database without opening its UI, but it was only used by my tests and didn't want to waste time figuring out how to speed it up. Now I have Linq2Acad, which has a clean example of how to import a block from another drawing in one readable line, so I replaced old code with one line of Linq2Acad. It's likely that using the middleman Linq2Acad I am going through a few more steps, just like Python does without strict typing, but now my code is more readable and faster.

End of rant (sorry, it just came out)

**Reply by wtertinek** [reply]
stenci,

unfortunately, there is no one rule on how to handle `IDisposable`s. You're right, you have to know what you're doing. I think [this post](https://www.reddit.com/r/csharp/comments/17jubuw/comment/k74217e/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) sums it up well:

> So really IDisposable is no different from any other form of memory management: whoever creates an object is supposed to document if they will dispose it. Something that uses it is supposed to claim responsibility. You have to be the person to examine the situation and decide which "something" is responsible.

Regarding the second part of your message: I'm very pleased that Linq2Acad is helping you to make your code more readable and faster. What helped me with larger AutoCAD C# projects is to build the project according to Ports & Adapters architecture. I once sketched an example [here](https://github.com/wtertinek/Linq2Acad/issues/33#issuecomment-1421698594). This gives you a very good separation between your business logic and the AutoCAD API stuff. Perhaps this will help you to implement your new project more easily.

Or maybe [CadPythonShell](https://github.com/chuongmep/CadPythonShell) can help you stick with Python at all?

Don't be sad, keep on trying. 😉

-Wolfgang

**Comment by stenci** [comment]
> Instead, something like a Reuse method would be much better which returns an instance of a new class, something like NotDisposableAcadDatabase:
>
> public static NotDisposableAcadDatabase Reuse(Transaction transaction)
> {
> }
> [...]
> If you want to further develop the idea and maybe also want to implement it, don'T hesitate to provide a pull request.

I tried to play a little with it, and I see two approaches: 

1.  ### Add a factory method that returns a `NotDisposableAcadDatabase` to the list of existing factory methods in `AcadDatabase`

    I tried to play a little with it, and, here are a few considerations:
    - One factory method creates an `AcadDatabase` from an existing `Transaction` is called `Active`
    - Two factory methods create `AcadDatabase`s from an existing `DataBase` are called `Use`
    
    For consistency, I would rather make the new factory method an overload, better a replacement, of the existing `Active`, than a new `ReUse`.
    
    On one hand this feels more consistent: `Active` creates an `AcadDatabase`(-ish) object out of a transaction.
    On the other hand, we have a seemingly factory method in `AcadDatabase` that returns a `NonDisposableAcadDatabase`.
    Soo... can we still call this a factory? A static method of a class that pretends to be a factory method and returns an instance of another class?

2.  ### Use a factory method / constructor of `NotDisposableAcadDatabase`

    An cleaner alternative would be to use the constructor of `NonDisposableAcadDatabase`:
    ```c#
    public NotDisposableAcadDatabase(Transaction transaction)
    {
      this.db = AcadDatabase.Active(transaction, false, false);
    }
    ```
    
    Used with:
    ```c#
    var db = new NotDisposableAcadDatabase(DwgManager.TopTransaction);
    ```
    
    One thing that I don't like with this solution would be that `AcadDatabase.Active(transaction, false, false)` still exists.

---

I have implemented the second option, the one that uses the constructor of `NotDisposableAcadDatabase` on [my fork](https://github.com/stenci/Linq2Acad).

The `NotDisposableAcadDatabase` class is a wrapper of `AcadDatabase` and exposes all its members with the exception of the factory members.

Do you think it's worth proceeding?

I noticed you have many constructors, well, factory members, I guess designed to work in different scenarios.
Do you think also `NotDisposableAcadDatabase` should have the same range of constructors / factory members?
Why did you go the factory rather than the constructor way?

**Reply by wtertinek** [reply]
Hi stenci, thanks for your input, I really appreciate it. You are right, there are four different scenarios:

Active: The database of the active document
Create: Create a new in-memory database
Open: Open a database from a file (ReadOny or ForWrite)
Use: Use an already opened database

Here, factory methods have been used to better name the scenarios.

I like your idea, a small extension comes to mind: what if we create a base class (`AcadDatabaseCore`) instead of the wrapper class? This maps the basic functionality of `AcadDatabase`, without any "life cycle management". I would remove the two overloads `Active(Transaction, bool, bool)` and `Use(Database, Transaction, boo, bool)` from `AcadDatabase`, instead `AcadDatabaseCore` would get the factory methods `Active(Transaction)` and `Use(Database, Transaction)`.

Since `AcadDatabase` now derives from `AcadDatabaseCore`, this would have the advantage that you can always pass on an `AcadDatabaseCore` to submodules of your own code that always comes without "life cycle management" (and so the client code cannot do anything wrong with it).

As an example:

```csharp
public class SubModule
{
   public SubModule(AcadDatabaseCore db)
   {
   }
}

// Scenario A
using (var db = AcadDatabase.Active())
{
   var module = new SubModule(db);
}

// Scenario B
var tr = GetTransactionFromSomewhere();
var db = AcadDatabaseCore.Active(tr);
var module = new SubModule(db);
```

For reasons of consistency, I would stick with factory methods in `AcadDatabaseCore`. I made a quick implementation [new_factory_method](https://github.com/wtertinek/Linq2Acad/compare/master...new_factory_method) branch.

What do you think?

**Comment by stenci** [comment]
The way I did it didn't feel right: having a class that can manage the transaction, and then a wrapper that uses it asking not to manage the transaction. 

I wanted to invert it, just like you did, that is make a class that does the job without managing the transaction and the _wrapper_ that derives from it and adds the transaction management, but I was afraid to break something. You have removed some factory methods, which is a breaking change, but it's the very breaking change we were looking for.

I also didn't like the name `NotDisposableAcadDatabase`. The name of a class should tell you what the class does, not what's wrong with its alternative.

I also didn't like to use a constructor next to your factory methods. I did it because it was easy and because I had not understood the semantic of your factory naming.

---

I have rebuilt the dll with your changes and used in my addin, and everything works well. 

I'm not (yet) using `Linq2Acad` extensively, but it already feels cleaner.

Now that the C# world and its disposables are finally making sense in my head, and now that I see your approach about hiding it when unnecessary and exposing it only when required, I'm starting to rethink some of my own classes.

Thank you for this!!

---

I had not understood the four scenarios. Seeing them in a list helps understanding that they exist and when to use them. Perhaps adding this would improve the documentation (I don't know how you are creating the documentation files, I'm new to C#, and my first and only attempt at it was to use Sandcastle, which uses the `NamespaceDoc` convention to add documentation for the namespace):
```c#
namespace Linq2Acad
{
  /// <summary>
  /// The classes <c>AcadDatabaseCore</c> and <c>AcadDatabase</c> contain factory methods to manage four scenarios:
  /// <list type="bullet">
  /// <item><description>Active: The database of the active document</description></item>
  /// <item><description>Create: Create a new in-memory database</description></item>
  /// <item><description>Open: Open a database from a file (ReadOny or ForWrite)</description></item>
  /// <item><description>Use: Use an already opened database</description></item>
  /// </list>
  /// These classes provide similar functionalities, but the general recommendation is:
  /// <list type="bullet">
  /// <item><description>AcadDatabaseCore: when the lifecycle of the transaction is managed outside</description></item>
  /// <item><description>AcadDatabase: to have automatic creation of the transaction and commit of the changes</description></item>
  /// </list>
  /// </summary>
  public static class NamespaceDoc
  {
    /* dummy class used for docstrings only */
  }

  /// <summary>
  /// The main class that provides access to the drawing database.
  /// </summary>
  public class AcadDatabaseCore
```

**Reply by wtertinek** [reply]
Thank you, your comment was very helpful. I didn't realise that the four scenarios are not clearly recognisable. This should also go straight into the "Getting started" section. Thanks for the code documentation, yes I use Sandcastle, I will add it.

Another small but I think important thing would be to rename `AcadDatabaseCore` to `AcadDataModel`. I think this will make things even clearer: we now have a separation between the "data source" / database (that handles the life cycle) and the actual Linq2Acad API, the data model.

So the updated sample would be:

```csharp
public class SubModule
{
   public SubModule(AcadDataModel dm)
   {
   }
}

// Scenario A
using (var db = AcadDatabase.Active())
{
   var module = new SubModule(db);
}

// Scenario B
var tr = GetTransactionFromSomewhere();
var dm = AcadDataModel.Active(tr);
var module = new SubModule(dm);
```

The only remaining question for me is whether it is clearly recognisable that `AcadDatabase` is an `AcadDataModel`?

**Reply by stenci** [reply]
Here is my 2 cents about class naming.

I like `AcadDatabase` because it's telling you what it is, but I don't like that it's [taken](https://help.autodesk.com/view/OARX/2023/ENU/?guid=OARX-ManagedRefGuide-Autodesk_AutoCAD_DatabaseServices_Database_AcadDatabase), and it could be confusing.

Perhaps `AcadDataModel` or `AcadDataLinq` both carry a similar meaning. This library doesn't need to be used with Linq, but it has Linq in its name, so `AcadDataLinq` sounds best to me.

So, if it is not too late to change the main class name (because it's already referred in the awesome blog), what about:
- `AcadDataModel`
- `AcadDataLinq` (this takes my vote)
- `AcadLinq`

and
- `AcadDataModelController`
- `AcadDataManager`
- `AcadDataLinqManager` (this takes my vote)

If you want a solution that takes care of the burocracy, you use `AcadDataLinqManager`.
If you want to manage the transaction yourself, you use `AcadDataLinq`.

**Reply by wtertinek** [reply]
Thanks for your suggestions stenci.

Some background: since the early days of LINQ, there has been the "LINQ to *" naming scheme for LINQ implementations: LINQ to Objects, LINQ to XML, there was LINQ to SQL, LINQ to NHibernate, Linq to Excel, etc. Following this tradition, I have named `Linq2Acad` just that.

I shy away from using `Linq` in a class name because it is non-idiomatic. `Linq` is always only part of a namespace or the name of a library, but not part of an identifier.

LINQ is the basic framework (via `IEnumerable<T>` or `IQueryable`) and is only visible in the API via the standard query operators, but the implementation is always bound to a different technology and therefore requires an entry point into the API. With LINQ to XML, for example, the entry point is `XElement`, with Entity Framework the context, with `Linq2Acad` the database. Since `Database` is too general and already in use, I originally decided in favour of `AcadDatabase`, because it is a good representation of the entry point to the AutoCAD database. I knew that `AcadDatabase` is already used as a (COM) property in the `Database` class and thought that this could be neglected because the COM interface is probably used very little in new AutoCAD versions (I might be wrong about that).

So much for the previous decision in favour of `AcadDatabase`.

Another factor is that it would be a fundamental change to the API to rename the existing main class. There would have to be a very good reason for this that would result in a significant improvement. We are not making an improvement for the `AcadDatabase` class, so in my opinion the effort is not worth it.

So I think I will stick with `AcadDatabase` and `AcadDataModel`. However, I would no longer bind the creation of `AcadDataModel` to the factory methods of `AcadDatabase`, but only use one single constructor `AcadDataModel(Database, Transaction)` that covers all cases:

```csharp
public class SubModule
{
   public SubModule(AcadDataModel dm)
   {
   }
}

// Scenario A
using (var db = AcadDatabase.Active())
{
   var module = new SubModule(db);
}

// Scenario B
var (db, tr) = GetContext();
var dm = new AcadDataModel(db, tr);
var module = new SubModule(dm);
```

And the fact that `AcadDatabase` can be used as `AcadDataModel` must be stated prominently in the documentation.

Thanks for your support and for thinking this through stenci.

**Reply by stenci** [reply]
Thanks for the background, all good info for me!

**Comment by kdub-nz** [comment]


**Reply by wtertinek** [reply]
Hi Kerry, I haven't had a chance to look at it yet, but I will in the next two weeks or so. This could become a new blog post...
